model/task/db.go

```diff
@@ -292,11 +298,15 @@ var (
 	AddAnnotations = []bson.M{
 		{
 			"$facet": bson.M{
-				// We skip annotation lookup for non-failed tasks, because these can't have annotations
+				// We skip annotation lookup for non-failed tasks, because these can't have annotations,
+				// and lookup is not necessary for tasks that already have HasAnnotationsKey set
 				"not_failed": []bson.M{
 					{
 						"$match": bson.M{
-							StatusKey: bson.M{"$nin": evergreen.TaskFailureStatuses},
+							"$or": []bson.M{
+								{HasAnnotationsKey: true},
+								{StatusKey: bson.M{"$nin": evergreen.TaskFailureStatuses}},
```

Comment:
Once we remove the backwards compatibility, we should remove this because we'll only ever have to look it up if it's set to true (and it'll already only be set to true if it's failed) . It may be a good idea to add a todo comment so we don't forget.

Reply:
Sounds good - added